clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -455,7 +456,13 @@ private static int parseAcks(String acksString) {
                         " to class " + producerConfig.getClass(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG).getName() +
                         " specified in value.serializer");
             }
-            int partition = partition(record, serializedKey, serializedValue, metadata.fetch());
+            // If topic is deleted between the start of this method when metadata was obtained
+            // and now, fail the send request.
+            Cluster cluster = metadata.fetch();
```

Comment:
Would it be better to just hold on to the `Cluster` object obtained at the start of this and use it throughout the call? I'm not sure how practical that is given the current code, but trying to handle inconsistencies in this data throughout the call seems like it'll be messy. For example, we accumulate messages and send based on topic partitions, does that also break if metadata changes? And do the issues highlighted here get even messier if the # of topic partitions change instead of the topic being entirely deleted?